### PR TITLE
Clean up old aggregation functions (Step 2)

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Move all server side code to use new aggregation functions.
1. Rewrite the IntermediateResultsBlock to use new aggregation functions.
2. Rewrite the Aggregation Operator and Combine Operator

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java [hunk 1]
```diff
@@ -138,35 +140,34 @@ public Block getNextBlock() {
...
 
+    List<AggregationInfo> aggregationInfos = _brokerRequest.getAggregationsInfo();
+    final AggregationFunctionContext[] aggregationFunctionContexts =
+        AggregationFunctionUtils.getAggregationFunctionContexts(aggregationInfos, null);
+    final int numAggregationFunctions = aggregationFunctionContexts.length;
+
     for (int i = 0; i < numOperators; i++) {
       final int index = i;
 
       _executorService.execute(new TraceRunnable() {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: can you extract new TraceRunnable() anonymous class to a separate class SegmentExecutableTask or something like that?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java [hunk 2]
```diff
@@ -138,35 +140,34 @@ public Block getNextBlock() {
...
           try {
-            blocks[index] = (IntermediateResultsBlock) _operators.get(index).nextBlock();
-            groupByResult = blocks[index].getAggregationGroupByResult();
+            IntermediateResultsBlock intermediateResultsBlock =
+                (IntermediateResultsBlock) _operators.get(index).nextBlock();
+            intermediateResultsBlocks[index] = intermediateResultsBlock;
+            aggregationGroupByResult = intermediateResultsBlock.getAggregationGroupByResult();
 
-            if (groupByResult != null) {
+            if (aggregationGroupByResult != null) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: when will this be null? where is the else block

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java [hunk 3]
```diff
@@ -49,12 +49,14 @@
 public class MCombineGroupByOperator extends BaseOperator {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: do we have multiple combine operators? i.e. group by, aggregation only, selection?. There is a lot of boilerplate code that can be extracted to one class

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineOperator.java [hunk 4]
```diff
@@ -182,12 +179,6 @@ public IntermediateResultsBlock callJob()
           new IntermediateResultsBlock(QueryException.getException(QueryException.EXECUTION_TIMEOUT_ERROR, e));
     }
 
-    // Trim merged results.
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineOperator.java]: we dont need this anymore?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java [hunk 5]
```diff
@@ -97,15 +69,12 @@ public Block getNextBlock(BlockId blockId) {
...
-    return new IntermediateResultsBlock(AggregationFunctionFactory.getAggregationFunction(_aggregationInfoList),
-        aggregationResults);
+    return new IntermediateResultsBlock(_aggregationFunctionContexts, aggregationExecutor.getResult(), false);
   }
 
   @Override
-  public boolean open() {
-    _projectionOperator.open();
-    return true;
+  public Block getNextBlock(BlockId blockId) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: better to keep the behavior symmetric across all operators, just check for blockId = 0 and return null for everything else?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/MinMaxRangeAggregationFunction.java [hunk 6]
```diff
@@ -162,7 +168,7 @@ public MinMaxRangePair merge(@Nonnull MinMaxRangePair intermediateResult1,
   @Nonnull
   @Override
   public FieldSpec.DataType getIntermediateResultDataType() {
-    return FieldSpec.DataType.DOUBLE;
+    return FieldSpec.DataType.OBJECT;
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/MinMaxRangeAggregationFunction.java]: will this cause an issue during server/broker upgrade

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java (hunk 3 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: We have 2 combine operators, one for aggregation only and selection, the other for aggregation group-by. I have a plan to merge them into a single operator (there is a TODO on that), and will address in another PR.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java [hunk 7]
```diff
@@ -138,35 +140,34 @@ public Block getNextBlock() {
...
-
-    // TODO: use new aggregation functions to combine results.
-    final List<AggregationFunction> aggregationFunctions =
-        AggregationFunctionFactory.getAggregationFunction(_brokerRequest);
-
     final Map<String, Serializable[]> resultsMap = new ConcurrentHashMap<>();
 
+    List<AggregationInfo> aggregationInfos = _brokerRequest.getAggregationsInfo();
+    final AggregationFunctionContext[] aggregationFunctionContexts =
+        AggregationFunctionUtils.getAggregationFunctionContexts(aggregationInfos, null);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: The second argument is the segment metadata. On per segment operator, we use segment metadata to construct a visitor and visit each aggregation function (visitor mode).

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java (hunk 1 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: I want to revisit this part in the combine operator clean up. The new group-by combine operator might not be ideal because it could occupy all threads available and block other queries. For this TraceRunnable, IMO, we'd better use a Callable and implement early termination when timeout happened. Will address in another PR.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java (hunk 2 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: When GroupByOperator hits some exceptions, this will be null. We don't need an else block here because we merge all exceptions anyway.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineOperator.java (hunk 4 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineOperator.java]: This trim only apply to group-by, which will never get into MCombineOperator.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/MinMaxRangeAggregationFunction.java (hunk 6 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/MinMaxRangeAggregationFunction.java]: This is a typo in the new added code (in clean up step 1), and the unit test caught it.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java (hunk 5 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: Actually we never allow calling getNextBlock with a blockId (the only exception is DataSource). Operators that allow calling getNextBlock() multiple times are BReusableFilterDocIdSetOperator and MProjectionOperator, and both of them will throw exception when getting called with a blockId. So to get symmetric behavior across all operators, the easier way is never call getNextBlock with a BlockId.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java (hunk 1 above)
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java]: ok. Makes sense to do that as part of a separate PR

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java (hunk 5 above)
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/AggregationOperator.java]: Agreed!
